File: packages/snow_draw_rust_engine/src/draw/elements/types/text/text_task_encoder.rs
#![allow(dead_code)]

use std::sync::Arc;

use crate::draw::elements::core::typed_element_render_task_encoder::{
    ElementState, RenderTaskList, TypedElementRenderTaskEncoder,
};
use crate::draw::types::edit_context::TextMetricsService;

use super::text_data::TextData;

/// Element-level render task for text payloads.
///
/// Mirrors Dart `TextRenderTask(element: ..., data: ..., localeTag: ...)`.
#[derive(Debug, Clone)]
pub struct TextRenderTask {
    pub element: ElementState,
    pub data: TextData,
    pub locale_tag: Option<String>,
}

impl TextRenderTask {
    /// Creates a new text render task snapshot.
    pub fn new(element: ElementState, data: TextData, locale_tag: Option<String>) -> Self {
        Self {
            element,
            data,
            locale_tag,
        }
    }
}

/// Encodes text elements into high-level render tasks.
///
/// Translation of Dart
/// `TextTaskEncoder extends TypedElementRenderTaskEncoder<TextData>`.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct TextTaskEncoder;

impl TextTaskEncoder {
    /// Creates a text task encoder.
    pub const fn new() -> Self {
        Self
    }
}

impl TypedElementRenderTaskEncoder<TextData> for TextTaskEncoder {
    fn encode_typed_tasks(
        &self,
        element: &ElementState,
        data: &TextData,
        locale_tag: Option<&str>,
        _text_metrics_service: Option<Arc<dyn TextMetricsService>>,
    ) -> RenderTaskList {
        vec![Box::new(TextRenderTask::new(
            element.clone(),
            data.clone(),
            locale_tag.map(str::to_owned),
        ))]
    }
}

File: packages/snow_draw_engine/lib/draw/elements/types/text/text_task_encoder.dart
import '../../../models/element_state.dart';
import '../../../render/tasks/render_tasks.dart';
import '../../../services/text/text_metrics_service.dart';
import '../../core/typed_element_render_task_encoder.dart';
import 'text_data.dart';

/// Encodes text elements into high-level render tasks.
final class TextTaskEncoder extends TypedElementRenderTaskEncoder<TextData> {
  /// Creates a text task encoder.
  const TextTaskEncoder();

  @override
  List<RenderTask> encodeTypedTasks({
    required ElementState element,
    required TextData data,
    String? localeTag,
    TextMetricsService? textMetricsService,
  }) => <RenderTask>[
    TextRenderTask(element: element, data: data, localeTag: localeTag),
  ];
}
